---
title: "Illustration of dreamlet on subset of NPS/AD"
subtitle: 'STAR-solo R1-4 merged h5ad'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{Decorrelate}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!--- 

# results 
# /sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/demo/results
cd /sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/
ml python git
git pull
R

system("git pull")
rmarkdown::render("demo_dreamlet.Rmd");


# https://hoffmg01.u.hpc.mssm.edu/nps_ad/demo_dreamlet.html


ml git
cd ~/build2/dreamlet
git pull
R CMD INSTALL .


--->


* Analyses to implement
  + Concordance as a function of 
    + number donors
    + number of cells
    + mixed model
    + batch effect
  + which genes have batch effect?
 + compare cell tpyes  Demonstrate performance by sampling two populations from the same cell cluster, and running showing the unpaired doesnt control false positive rate.  How to demonstrate false positive rate.
  


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```

```{r load.packages, cache=FALSE}
# Use cache=FALSE so that package are fully loaded each time
# This ensures that forks within mclapply() have these loaded
# Othewise, mclapply() not have access to these libraries and will fail 
#   unless the libraries are manually loaded within each fork
suppressPackageStartupMessages({
library(SingleCellExperiment)
library(zellkonverter)
library(DelayedArray)
library(HDF5Array)
library(GSEABase)
library(dreamlet)
library(muscat)
library(cowplot)
library(zenith)
library(scater)
library(data.table)
library(S4Vectors)
library(tidyverse)
library(kableExtra)
library(qvalue)
library(corrplot)
library(org.Hs.eg.db)
})
```


```{r load.data, cache=FALSE}
outfolder = '/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/demo/results/'

datafile = paste(outfolder, "demo_dreamlet_sce.RDS")

if( file.exists(datafile) ){
  # reading data from RDS is much faster:
  #   especially good for pipeline development
  sceCombine = readRDS( datafile )
}else{
  # read snRNA-seq data with annotations build in
  h5ad_file = "/sc/arion/projects/CommonMind/leed62/NPS-AD/analysis_solo/NPS-AD_roundMerged_pass2_anno.h5ad"

  # read raw/* from h5ad file
  sce_in = readH5AD(h5ad_file, use_hdf5=TRUE, raw=TRUE, verbose=TRUE)

  # extract raw counts
  sceCombine = altExp( sce_in, "raw", withDimnames=TRUE, withColData=TRUE)

  sceCombine = as(sceCombine, "SingleCellExperiment")
  reducedDims(sceCombine) <- reducedDims(sce_in)

  # set metadata to empty to reduce size and speep up development time
  metadata(sceCombine) = list()

  # collapse GABAergic neurons
  sceCombine$celltype8 = sceCombine$celltype %>%
    as.character %>% 
    recode( 
    "GABAergic SST interneuron" = 'GABAergic neuron',
    "GABAergic VIP interneuron" = 'GABAergic neuron',
    "GABAergic PVALB interneuron" = 'GABAergic neuron') %>% 
    as.factor

  saveRDS(sceCombine, file=datafile)
}
```

Data freeze includes `r length(table(sceCombine$Channel))` samples, `r length(table(sceCombine$round_num))` rounds, `r length(table(sceCombine$batch))` 10X batches, `r length(table(sceCombine$SubID))` donors, and `r format(ncol(sceCombine), big.mark=',')` cells passing QC.

# Joint UMAP
```{r umap2, dev="png"}
# extract UMAP coordinates and annotations
df = cbind(reducedDim(sceCombine, "X_umap"), colData(sceCombine)[,c("anno", "celltype", 'celltype8', "class", "subtype", "leiden_labels")])
df = data.frame(df)

ggplot(df, aes(V1, V2, color=celltype)) + geom_point(size=.05) + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2")
```

```{r umap3, dev="png"}
ggplot(df, aes(V1, V2, color=class)) + geom_point(size=.05) + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2")
```

```{r umap8, dev="png"}
ggplot(df, aes(V1, V2, color=celltype8)) + geom_point(size=.05) + theme_classic() + theme(aspect.ratio=1) + guides(colour = guide_legend(override.aes = list(size = 1.5))) + xlab("UMAP1") + ylab("UMAP2")
```

```{r combineData}
# Specify how to collapse into pseudo-bulk
sceCombine$id <- paste0(sceCombine$batch, '_', sceCombine$SubID)

# Create pseudo-bulk SingleCellExperiment
pbObj <- aggregateToPseudoBulk(sceCombine,
    assay = "X", 
    fun = "sum",
    cluster_id = "celltype8",
    sample_id  = "id")

# Combine with phenotype data at the Donor level
# read metadata
meta_file = "/sc/arion/projects/CommonMind/leed62/NPS-AD/MSSM_HBCC_RUSH_clinical_metadata_combo.csv"
df_meta = fread(meta_file)
df_meta = df_meta[,-1]

# save original order of samples
colData(pbObj)$row.order = 1:nrow(colData(pbObj))

# g = intersect(colnames(colData(pbObj)), colnames(df_meta))

# merge colData with additional metadata
df_merge = merge(colData(pbObj), df_meta, by="SubID", all.x=TRUE)

# set rownames to be same as original, based on row.order
# These rownames determine colnames(pbObj), which is used downstream
rownames(df_merge) = rownames(colData(pbObj))[df_merge$row.order]

# extract batch info per sample
# df_batching = with( colData(sceCombine), unique(data.frame(id, round_num, batch, prep, HTO)))
# df_merge = merge(df_merge, df_batching, by.x= "row.names", by.y="id", all.x=TRUE)

# rownames(df_merge) = df_merge$Row.names
# df_merge = df_merge[,-1]

# save merged data as colData while retaining original cell order
colData(pbObj) = df_merge[order(df_merge$row.order),]
colData(pbObj)$row.order = c()# remove row.order

# Set levels for Dx2
lvls = sort(unique(as.character(colData(pbObj)$Dx2)))
lvls = lvls[-which(lvls=="Control")]
colData(pbObj)$Dx2 = factor(colData(pbObj)$Dx2, c("Control", lvls))

saveRDS(pbObj, file="/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/demo_dreamlet_pbobj.RDS" )
```

# Analysis of joint dataset

```{r round_num.summary}
table(pbObj$round_num)
```

```{r Dx.summary}
df = unique(data.frame(Dx2 = pbObj$Dx2, Donor = pbObj$SubID))

sort(table(df$Dx2), decreasing=TRUE) %>% kbl() %>% kable_styling(full_width=FALSE)
```

```{r voom}
# Normalize and apply voom
res.proc = processAssays( pbObj, ~ (1|SubID) + (1|batch) + (1|round_num) + (1|HTO) + (1|cohort), 
  min.cells = 5,
  min.count = 10, 
  # pmetadata = df_collapse,  
  # pkeys=c('ID', 'anno'),
  BPPARAM = SnowParam(16, progressbar=TRUE))
  
res.proc
```

```{r voom.plot, fig.height=10, fig.width=10, cache=TRUE}
plotVoom( res.proc )
```


```{r voom.plot_sub, fig.height=4, fig.width=10, cache=TRUE}
plotVoom( res.proc[7:9] )
```



```{r varPart}
form = ~ (1|SubID) + (1|batch) + (1|round_num) + (1|HTO) + (1|Dx2) + (1|Sex) + (1|cohort) + scale(Age) + scale(PMI) 

res.vp = fitVarPart(res.proc, form, BPPARAM = SnowParam(18, progressbar=TRUE))
```

```{r vp.plot, fig.height=15, fig.width=10, cache=TRUE}
colnames(res.vp) = gsub("scale\\.", "", colnames(res.vp))
colnames(res.vp) = gsub("\\.", "", colnames(res.vp))
colnames(res.vp) = gsub("SubID", "Donor", colnames(res.vp))

plotVarPart( sortCols(res.vp), label.angle=45 )
```

```{r vp.plot_sub, fig.height=4, fig.width=10, cache=TRUE, eval=FALSE}
idx = res.vp$assay %in% c('GABAergic neuron', 'Glutamatergic neuron', 'Microglia')
plotVarPart( sortCols(res.vp[idx,]), label.angle=70 )
```

### Examine variancePartition results
```{r vp.examine, fig.height=3, fig.width=7}
CT = "Astrocyte"

i = 1:3
df = sortCols(res.vp)[res.vp$assay == CT,]
rownames(df) = df$gene
i = c(i, which.max(df$Donor), which.max(df$Batch), which.max(df$Residuals))

plotPercentBars( data.frame(df[i,-c(1,2)])) + ggtitle(CT)
```

Examples

```{r vp.examine2, fig.height=6, fig.width=8, cache=TRUE}      
figList = lapply( df$gene[i], function(geneName){
  df = merge( as.data.frame(t(assay(res.proc, CT)$E[geneName,,drop=FALSE])), colData(res.proc), by="row.names")

  plotStratifyBy(df, 'SubID', geneName, colorBy=NULL, main=geneName) + theme(aspect.ratio=1)
})
plot_grid( plotlist=figList, nrow=2 )
```

## Batch effects and gen sets
```{r DE.batch}
library(zenith)
go.gs = get_GeneOntology(to="ENSEMBL")

coef = "batch"
n_genes_min = 10
geneSets = go.gs

res = lapply( unique(res.vp$assay), function(key){
  
  # subset by assay
  df = res.vp[res.vp$assay == key,]

  # get variance statistic
  statistic = df[,coef]
  names(statistic) = df$gene

  # convert GeneSetCollection to list
  geneSets.lst = recodeToList( geneSets )

  # Map from Ensembl genes in geneSets_GO to 
  # from trimmed Ensembl names from RNA-seq data 
  index = ids2indices( geneSets.lst, names(statistic))
     
  # filter by size of gene set
  index = index[vapply(index, length, FUN.VALUE=numeric(1)) >= n_genes_min]

  res = cameraPR( statistic, index, use.ranks=TRUE)

  DataFrame( assay = key, Geneset = rownames(res), res)
})
res = do.call(rbind, res)

res$FDR = p.adjust(res$PValue, "fdr")

# table(res$FDR < 0.05)
```


## Correlation with and between donors
```{r within_btw, fig.height=10, fig.width=10}
source("/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/common_analyses.R")

eval_within_across_donor( res.proc, "SubID" )
```

## dreamlet analysis
```{r dreamlet.fit}
form = ~ Sex + (1|SubID) + (1|batch) + (1|round_num) + (1|HTO) + (1|cohort) + Dx2 + scale(Age) + scale(PMI)

fit.dl = dreamlet( res.proc, form, BPPARAM=SnowParam(36))

saveRDS( fit.dl, "/sc/arion/projects/CommonMind/hoffman/NPS-AD/work/fit.dl.RDS")
```



```{r de.summary}
coefs = c('Dx2AD', 'Dx2SCZ')

tabLst = lapply( coefs, function(key){
  tab = topTable(fit.dl, coef=key, number=Inf)

  tab %>% 
    as_tibble %>% 
    group_by(assay) %>% 
    summarise( coef = key, 
      nGenes = length(adj.P.Val), 
      nDE = sum(adj.P.Val < 0.05),
      pi1 = 1 - qvalue(P.Value)$pi0)
  })
tab = do.call(rbind, tabLst)

as.data.frame(tab)
```

```{r de.plots}
ggplot(tab, aes(assay, nGenes, fill=assay)) + geom_bar(stat="identity") + facet_wrap(~coef) + theme_classic() + theme(aspect.ratio=1, legend.position="none", axis.text.x = element_text(angle = 45, vjust=1, hjust=1) )

ggplot(tab, aes(assay, nDE, fill=assay)) + geom_bar(stat="identity") + facet_wrap(~coef) + theme_classic() + theme(aspect.ratio=1, legend.position="none", axis.text.x = element_text(angle = 45, vjust=1, hjust=1) )

ggplot(tab, aes(assay, pi1, fill=assay)) + geom_bar(stat="identity") + facet_wrap(~coef) + theme_classic() + theme(aspect.ratio=1, legend.position="none", axis.text.x = element_text(angle = 45, vjust=1, hjust=1) )
```

# Volcanos
```{r volcanos.1, fig.width=7, fig.height=12}
plotVolcano( fit.dl, coef = coefs[1], nGenes=0 ) + ggtitle(coefs[1])
```

```{r volcanos.2, fig.width=7, fig.height=12}
plotVolcano( fit.dl, coef = coefs[2], nGenes=0 ) + ggtitle(coefs[2])
```

# Cross-disease signature analysis
```{r compare.signatures, fig.width=9, fig.height=9}
# get one data.frame with all differential expression scores
df = lapply( coefs, function(coef){
  data.frame( coef=coef, topTable(fit.dl, coef=coef, number=Inf))
})
df = do.call(rbind, df)

# dcast as wide matrix
df2 = reshape2::dcast(df, ID ~ coef + assay , value.var="z.std")

colnames(df2) = gsub("^Dx2", '', colnames(df2))
C = cor(df2[,-1], use='pairwise.complete.obs')

col = colorRampPalette(c("blue3", "white","red3"))(200)

corrplot(C, col=col, tl.col="black")

corrplot(C, order="hclust", col=col, tl.col="black")
```



# zenith for gene set analysis
```{r zenith}
# Load Gene Ontology database 
go.gs = get_GeneOntology()

ngenes = sapply(go.gs, function(gs) length(geneIds(gs)))
go.gs = go.gs[(ngenes > 20) & (ngenes < 2000)]

coefs = c('Dx2AD', 'Dx2SCZ')

res.gsa = lapply( coefs, function(key){

  # check inter.gene.cor=NA
  zenith_gsa(fit.dl, coefs=key, go.gs, inter.gene.cor=.01)
})
names(res.gsa) = coefs
res.gsa_cat = do.call(rbind, res.gsa)
```

```{r zenith.heatmap.1, fig.width=7, fig.height=12}
fig = plotZenithResults(res.gsa[[1]], 3, 1) + ggtitle(names(res.gsa)[1])
```

```{r zenith.heatmap.2, fig.width=7, fig.height=12}
fig = plotZenithResults(res.gsa[[2]], 3, 1) + ggtitle(names(res.gsa)[3])
```


# Split data
```{r split.data}
pbList = list( A = pbObj[,pbObj$round_num %in% 1:2],
               B = pbObj[,pbObj$round_num %in% 3:4])

# summarize disease status
df = lapply( names(pbList), function(x){
  pb = pbList[[x]]
  unique(data.frame(Set =x, Dx2 = pb$Dx2, Donor = pb$SubID))
  })
df = do.call(rbind, df)

df %>% 
  group_by(Set) %>% 
  count(Dx2, Set, sort = TRUE) %>%
  kbl() %>% kable_styling(full_width=FALSE)
```


```{r voom.2}
res.procList = lapply( pbList, function(pb){
 
  idx = which(pb$Dx2 %in% c("AD", "Control", "SCZ"))

  # Normalize and apply voom
  processAssays(pb[,idx], ~ (1|SubID) + (1|batch) + (1|HTO) + (1|cohort), 
    min.cells = 5,
    min.count = 10, 
    BPPARAM = SnowParam(16, progressbar=TRUE))
})   
```


```{r varPart.2}
form = ~ (1|SubID) + (1|batch) + (1|HTO) + (1|Dx2) + (1|Sex) + (1|cohort) + scale(Age) + scale(PMI)      

res.vpList = lapply( res.procList, function(x){
  fitVarPart(x, form, BPPARAM = SnowParam(12, progressbar=TRUE))
})
```

### Differential expression in each batch 
Compare 1) fixed effects collapsing model, 2) full pwLMM, 3) no batch correction
f(res.dlList) to evaluate concordance.  Correlation of Dx signal with GC content

```{r dreamlet.2}
form = ~ (1|SubID) + (1|batch) + (1|HTO) + (1|Dx2) + (1|Sex) + (1|cohort) + scale(Age) + scale(PMI) + Dx2
   
res.dlList = lapply( res.procList, function(x){
  dreamlet(x, form, BPPARAM = SnowParam(12, progressbar=TRUE))
}) 
```

### Concordance between sets
Correlation between t-statistics
```{r de.summary.2}
library(broom)
coefs = c('Dx2AD', 'Dx2SCZ')

resTab = lapply( coefs, function(key){

  tab = lapply( res.dlList, function(fit){
    tab = topTable(fit, coef=key, number=Inf)
    tab$key = with(tab, paste(assay, ID, key))
    data.frame(coef=key, tab)
  })
  tab = inner_join(tab[[1]], tab[[2]], by='key')

  tab %>%  
    as_tibble %>% 
    group_by(assay.x) %>% 
    group_modify(~ tidy(cor.test(.x$t.x, .x$t.y, method="spearman"))) %>% 
    mutate(coef = key)
  })
resTab = do.call(rbind, resTab)

ggplot(resTab, aes(assay.x, estimate, fill=assay.x)) + geom_bar(stat="identity") + theme_classic() + theme(aspect.ratio=1, legend.position="none") + facet_wrap(~coef) + coord_flip()
```

```{r plots}
library(ggplot2)
library(viridis)

# Compute two dimensional density
get_density <- function(x, y, n = 250) {
  dens <- MASS::kde2d(x = x, y = y, n = n)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

# make scatter plot with 2D density using viridis colors
plotScatterDensity = function(value1, value2){

  # convert two vectors to a data frame
  df = data.frame(cbind(value1, value2))

  # determine limits of the plot
  lim = with(df, max(abs(c(value1, value2))))
  
  # Compute 2D density
  df$density <- get_density(df$value1, df$value2, n = 100)
  
  # Scatter plot colored by density
  ggplot(df, aes(value1, value2, color=density)) + geom_point(size=.4) + theme_bw(16) + theme(aspect.ratio=1, legend.position="bottom", plot.title = element_text(hjust = 0.5)) + geom_abline(color="red") + geom_vline(xintercept=0, col="grey40", linetype="dashed") + geom_hline(yintercept=0, col="grey40", linetype="dashed")+ scale_color_viridis() + geom_smooth(method="lm", se=FALSE, color="darkorange")#  + xlim(-lim, lim) + ylim(-lim, lim) 
}
```

```{r concordance.t.stat}

tabA = topTable(res.dlList[['A']], coef='Dx2SCZ', number=Inf) %>%
        as_tibble  %>%
        filter(assay == "OPC") 

tabB = topTable(res.dlList[['B']], coef='Dx2SCZ', number=Inf) %>%
        as_tibble  %>%
        filter(assay == "OPC") 

df = inner_join(tabA, tabB, by="ID") 

plotScatterDensity(tabB$AveExpr, tabB$logFC)

#ggsave("test.png", fig)
```








```{r concordance}

get_pi1 = function(p){

  res <- tryCatch( pi0est(p), error = function(e) NULL)

  pi1 = NA
  if( ! is.null(res)){
    pi1 = 1 - res$pi0
  }
  pi1
}


# stats
coefs = c('Dx2AD', 'Dx2SCZ')

tab = lapply( names(res.dlList), function(key){
  tab = lapply(coefs, function(coef){
    tab = topTable(res.dlList[[key]], coef=coef, number=Inf)
    
    tab %>%
      as_tibble %>%
      group_by(assay) %>%
      summarize(  nGenes = length(adj.P.Val),
      nDE = sum(adj.P.Val < 0.05),
      pi1 = get_pi1(P.Value))  %>%
      mutate(key = key, coef = coef)
  })
  do.call(rbind, tab)
})
tab = do.call(rbind, tab)







resTab = lapply( coefs, function(key){

  tab = lapply( res.dlList, function(fit){
    tab = topTable(fit, coef=key, number=Inf)
    tab$key = with(tab, paste(assay, ID, key))
    data.frame(coef=key, tab)
  })
  tab = inner_join(tab[[1]], tab[[2]], by='key')

  tab %>%  
    as_tibble %>% 
    group_by(assay.x) %>% 
        summarise( 
        nGenes = length(adj.P.Val.x), 
        nDE = sum(adj.P.Val.x < 0.05),
        pi1.x = get_pi1(P.Value.x[adj.P.Val.y <= 0.05]),
        pi1.y = get_pi1(P.Value.y[adj.P.Val.x <= 0.05])) %>%
    mutate(coef = key)
  })
resTab = do.call(rbind, resTab)

```


### GC content and batch effects
```{r gc.content}
# Compute GC content
# adapted from https://github.com/dpryan79/Answers/blob/master/SEQanswers_42420/GTF2LengthGC.R
library(GenomicRanges)
library(rtracklayer)
library(Rsamtools)

# CellRanger
# GTFfile = "/hpc/packages/minerva-centos7/cellranger/5.0.0/refdata-cellranger/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
# FASTAfile = "/hpc/packages/minerva-centos7/cellranger/5.0.0/refdata-cellranger/refdata-gex-GRCh38-2020-A/fasta/genome.fa"

GTFfile = "/sc/arion/projects/psychAD/ref_GRCh38/Homo_sapiens.GRCh38.104.gtf"
FASTAfile = "/sc/arion/projects/psychAD/ref_GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa"

GTF <- import.gff(GTFfile, format="gtf", genome="GRCh38.104", feature.type="exon")
grl <- GenomicRanges::reduce(split(GTF, elementMetadata(GTF)$gene_id))
reducedGTF <- unlist(grl, use.names=T)
elementMetadata(reducedGTF)$gene_id <- rep(names(grl), elementNROWS(grl))

# Open the fasta file
FASTA <- FaFile(FASTAfile)
open(FASTA)

#Add the GC numbers
elementMetadata(reducedGTF)$nGCs <- letterFrequency(getSeq(FASTA, reducedGTF), "GC")[,1]
elementMetadata(reducedGTF)$widths <- width(reducedGTF)

#Create a list of the ensembl_id/GC/length
calc_GC_length <- function(x) {
    nGCs = sum(elementMetadata(x)$nGCs)
    width = sum(elementMetadata(x)$widths)
    c(width, nGCs/width)
}
df_GC <- t(sapply(split(reducedGTF, elementMetadata(reducedGTF)$gene_id), calc_GC_length))
colnames(df_GC) <- c("Length", "GC")
df_GC = as.data.frame(df_GC)
```

### Concordance with batch effect in each freeze
```{r examine, fig.width=8, fig.height=14}
library(viridis)
  
figList = lapply(res.vpList, function(vp){
  plotVarPart(sortCols(vp), label.angle=45)
  })

figList[[1]] + ggtitle("Freeze 1")
figList[[2]] + ggtitle("Freeze 2")
```

## Batch effects in freezes A and B
```{r compare, fig.width=7, fig.height=10}
df = lapply(names(res.vpList), function(id){
  df = data.frame(Set = id, res.vpList[[id]])  
  df$key = with(df, paste(gene, assay))
  df
  })
df = merge(df[[1]], df[[2]], by="key")

get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

key = "batch"
a = paste0(key, '.x')
b = paste0(key, '.y')

df$density <- get_density(df[[a]], df[[b]], n = 100)

ggplot(df, aes_string(a,b)) + geom_point(aes(color=density), size=.02) + facet_wrap(~assay.y, ncol=3) + theme_classic() + theme(aspect.ratio=1) + geom_abline(color="grey50", linetype="dashed", width=.3)  + scale_color_viridis() + xlab("Freeze A") + ylab("Freeze B")
```

```{r barplot}
df %>%
  group_by(assay.y) %>%
  summarize(Corr = cor(batch.x, batch.y, method="spearman")) %>%
  ggplot(aes(assay.y, Corr, fill=assay.y)) + 
    geom_bar(stat="identity") +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    coord_flip() +
    ylab("Spearman Correlation in batch effect across freezes") +
    xlab("Cell type") +  
    scale_fill_discrete(name = "Cell type") +
    scale_y_continuous(limits=c(0,NA), expand=c(0,0))
```

```{r corr}
library(broom)

df2 = merge(df, df_GC, by.x="gene.x", by.y="row.names")

df_res = lapply(colnames(df)[c(5:13, 17)], function(key){

  res = df2 %>% 
    as_tibble %>% 
    group_by(assay.y) %>%  
    group_modify(~ tidy(cor.test(.x[[key]], .x$GC, method="spearman"))) %>%
    left_join(df2 %>% group_by(assay.y) %>% tally, by="assay.y" ) %>%
    mutate(se = sqrt((1-estimate^2)/(n-2)))

  res$key = key
  res
})
df_res = do.call(rbind, df_res)
df_res$fdr = p.adjust(df_res$p.value, 'fdr')

df_res$key = gsub("scale", "", df_res$key)
df_res$key = gsub("\\.x", "", df_res$key)

ggplot(df_res, aes(assay.y, key, fill=estimate)) + geom_tile() + geom_text(data=subset(df_res, fdr < 0.05), aes(assay.y, key,label='*')) + scale_fill_gradient2(low="blue", mid="white", high="red", limits=c(-1, 1)) + theme_classic() + theme(axis.text.x = element_text(angle = 75, vjust = 1, hjust=1))
```



```{r barplot2}
df2 = merge(df, df_GC, by.x="gene.x", by.y="row.names")

df_res = lapply( c("batch.x", "batch.y"), function(key){

  res = df2 %>% 
    as_tibble %>% 
    group_by(assay.y) %>%  
    group_modify(~ tidy(cor.test(.x[[key]], .x$GC, method="spearman"))) %>%
    left_join(df2 %>% group_by(assay.y) %>% tally, by="assay.y" ) %>%
    mutate(se = sqrt((1-estimate^2)/(n-2)))

  res$key = key
  res
})
df_res = do.call(rbind, df_res)
df_res$fdr = p.adjust(df_res$p.value, 'fdr')

df_res %>%
  # filter(key == "batch") %>%
  mutate(fdr = p.adjust(p.value, "BH")) %>%
  group_by(assay.y) %>%
  ggplot(aes(assay.y, estimate, fill=assay.y, alpha=key)) + 
    geom_bar(stat="identity", position = position_dodge(width = 0.9)) +
    geom_errorbar(aes(ymin = pmax(0,estimate-1.96*se), ymax = estimate+1.96*se), width=0, position = position_dodge(width = 0.9)) +
    theme_classic() +
    theme(aspect.ratio=1, legend.position="none") +
    coord_flip() +
    ylab("Spearman correlation between batch effect and GC content per gene") +
    xlab("Cell type") +  
    scale_fill_discrete(name = "Cell type") +
    scale_y_continuous(limits=c(0,.31), expand=c(0,0)) +
    geom_text(aes(y=0.28, label=ifelse(fdr < 0.05, "            #", '')), 
      , position = position_dodge(width = 0.9)) +
    scale_alpha_discrete(range=c(1, .5))
```

```{r GC.batch, eval=FALSE}

fig = df2 %>% 
    as_tibble %>% 
    group_by(assay.y) %>%  
    filter(assay.y == "Glutamatergic neuron") %>%  
    ggplot(aes(GC, batch.y)) +
      geom_point()

df3 = df2 %>% 
    as_tibble %>% 
    group_by(assay.y) %>%  
    filter(assay.y == "Glutamatergic neuron")

fig = plotScatterDensity(df3$GC, rank(df3$batch.x)      )

with(df3, cor(GC, batch.x, method="sp"))

# ggsave("test.png", fig)

```




