---
title: "Testing/development for dreamlet/microglia"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
vignette: >
  %\VignetteIndexEntry{dreamlet/microglia}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---


<!--- 
cd /hpc/users/hoffmg01/work/nps_ad
ml python
R

system("ml git; git pull")
rmarkdown::render("microglia_test.Rmd");

# https://hoffmg01.u.hpc.mssm.edu/nps_ad/
--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  cache = TRUE,
  cache.lazy = FALSE)
```

```{r load.packages, cache=FALSE}
# Use cache=FALSE so that package are fully loaded each time
# This ensures that forks within mclapply() have these loaded
# Othewise, mclapply() not have access to these libraries and will fail 
#   unless the libraries are manually loaded within each fork
suppressPackageStartupMessages({
library(dreamlet)
library(zenith)
library(mashr)
})

# Must install mashr from GitHub (not CRAN!)
# remotes::install_github("stephenslab/mashr")
```

```{r basic.processing}

# set path
setwd("/sc/arion/projects/Microglia/scRNAseq_proc/analysis_dreamlet")

# read pseudobulk
pbObj = readRDS("210927_fresh-mg_pass2.export.pbObj.RDS")

# I only used fixed effects here for speed in development new ideas
# voom
res.proc = processAssays(pbObj, ~ Source + age + sex + dx, min.count=5)

# differential expression
res.dl = dreamlet(res.proc, ~ Source + age + sex + dx)
```

# Gene set enrichment using zenith
`zenith` defaults are *super* conservative.  Here I used `inter.gene.cor=0.01` which is the default in `limma::camera`.  This more liberal setting now find more differential gene sets, than the default settings. (Note, genesets are sorted by cell type then p-value)
```{r zenith}
go.gs = get_GeneOntology(to="SYMBOL")

resGS = zenith_gsa(res.dl, coef="dxAD", go.gs, inter.gene.cor=0.01)

head(resGS)
```

# [Mashr](https://www.nature.com/articles/s41588-018-0268-8) models high dimensional testing under many conditions.
Here we test genes in each cell type.  Standard analysis considers each cell type separately, but we can consider a joint modelling across cell types and genes to borrow information using a Bayesian framework:

```{r mashr}
# convert results table to matrix
toMatrix = function(tab, col){
  
  # row and column names
  rn = unique(tab$ID)
  cn = unique(tab$assay)

  i = match(tab$ID, rn)
  j = match(tab$assay, cn)

  M = sparseMatrix(i,j, x=tab[[col]], 
    dims=c(length(rn), length(cn)),
    dimnames = list(rn, cn))

  data = as.matrix(M)
  data[data == 0] = NA 
  data
}

# get results for each gene and cell type
tab = topTable(res.dl, coef="dxAD", Inf)

# compute standard error from t-stat and logFC
tab$se = tab$logFC / tab$t

# convert to matricies
B = toMatrix(tab, "logFC")
S = toMatrix(tab, "se")

# run mashr on these matricies
#-----------------------------

# set up
# NA's are replaced with beta = 0 with se = 1e6 
data = mash_set_data(B, S)

# estimate some parameters
U.c = cov_canonical(data)

# Estimate correlation structure 
V.em = mash_estimate_corr_em(data, U.c, details = TRUE)

# get model fit
m.Vem = V.em$mash.model
```



[Source](https://github.com/GabrielHoffman/nps_ad/blob/master/microglia_test.Rmd)





















