---
title: "Test pseudotime"
subtitle: 'Fresh Microglia and NPSAD'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
    toc: true
    smart: false
---


<!--- 

# cd /Users/gabrielhoffman/workspace/repos/nps_ad/mg

cd /sc/arion/projects/CommonMind/hoffman/NPS-AD/work/nps_ad/mg
R --vanilla

system("ml git; git pull"); rmarkdown::render("test_pseudotime.Rmd");


# https://hoffmg01.hpc.mssm.edu/mg/test_pseudotime.html


--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c("png", "pdf"),
  package.startup.message = FALSE,
  cache = TRUE,
  cache.lazy = FALSE)
```


```{r load.packages}
library(tidyverse)
library(variancePartition)
library(lmerTest)
library(ggplot2)
library(zellkonverter)
library(SingleCellExperiment)
```

```{r analysis}
# PsychAD MicroPVM + metadata
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/5_taxonomy/1_scANVI_psychAD_annotation/230420_PsychAD_MicroPVM_independent_meta_scANVI_new-umap.h5ad"
# sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE)
# saveRDS(colData(sce), file="/sc/arion/scratch/hoffmg01/data/colData.RDS")
sce_colData = readRDS("/sc/arion/scratch/hoffmg01/data/colData.RDS")

# Fresh MG
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/4_qc/230424_FreshMG_freeze3_MicroPVM_anno_meta_HGNC_log1p.h5ad"
# sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE, obsp=FALSE)
# saveRDS(colData(sce), file="/sc/arion/scratch/hoffmg01/data/colData_fresh.RDS")
sce_colData_fresh = readRDS("/sc/arion/scratch/hoffmg01/data/colData_fresh.RDS")

files = dir("/sc/arion/scratch/hoffmg01/data/", pattern="*csv.gz",recursive=TRUE, full.names=TRUE)
names(files) = gsub(".csv", "", paste(basename(dirname(files)), basename(files), sep="/"))

dfList = lapply( names(files)[1:2], function(id){

	df = read_csv(files[id])
	# df$SubID =  gsub("-[GCTA]+", "", df$barcodekey)

	df2 = merge(df[,1:3], sce_colData_fresh, by.x="barcodekey", by.y="row.names")
	as.data.frame(df2)

	})
names(dfList) = names(files)[1:2]

dfList2 = lapply( names(files)[3:4], function(id){

	df = read_csv(files[id])
	colnames(df)[1] = "idx"

	df = cbind(df[,1:3], sce_colData[df$idx+1,])

	df
	})
names(dfList2) = names(files)[3:4]

dfList = c(dfList, dfList2)
```

```{r plots, fig.width=12}
fitList = lapply(names(dfList), function(id){
	message(id)
	ggplot(dfList[[id]], aes(as.character(braak), dpt_pseudotime, fill=subtype)) +
	geom_violin() +
	theme_classic() +
	theme(aspect.ratio=1, legend.position="none") +
	xlab("Braak") + 
	ggtitle(id) +
	facet_wrap(~ subtype, nrow=2)
})
fitList
```


```{r lmer}
res = lapply(names(dfList), function(id){

	if( grepl("fresh_mg", id) ){
		form = dpt_pseudotime ~ braak + (1|subtype) + (1|Channel) + age + (1|sex) + PMI + log(n_genes) + (1|predicted_phase)
	}else{
		form = dpt_pseudotime ~ braak + (1|subtype) + (1|Channel) + age + (1|sex) + PMI + (1|poolID) + log(n_genes) + (1|predicted_phase)
	}

	data = dfList[[id]][,all.vars(form)]

	fit = lmer(form, data, REML=FALSE, control=lme4::lmerControl(optimizer="Nelder_Mead"))

	list(vp = calcVarPart(fit), summary = coef(summary(fit)))
})
names(res) = names(dfList)
```

```{r show.lmer, cache=FALSE}
res
```


```{r lmer.subtype}
res = lapply(names(dfList), function(id){

	data = dfList[[id]]

	res = lapply(unique(data$subtype), function(st){

		if( grepl("fresh_mg", id) ){
			form = dpt_pseudotime ~ braak + (1|Channel) + age + (1|sex) + PMI + log(n_genes) + (1|predicted_phase)
		}else{
			form = dpt_pseudotime ~ braak + (1|Channel) + age + (1|sex) + PMI + (1|poolID) + log(n_genes) #+ (1|predicted_phase)
		}

		include = data$subtype == st
		data2 = droplevels(data[include,all.vars(form)])

		fit = lmer(form, data2, control=lme4::lmerControl(optimizer="Nelder_Mead"))

		data.frame( Dataset = id, 
					subtype = st,
					t(coef(summary(fit))["braak",]))
	})
	do.call(rbind, res)
})
res = do.call(rbind, res)
```






# Test pseudotime
```{r testPT, eval=FALSE}
library(dreamlet)
library(MASS)
file = "/sc/arion/projects/Microglia/scRNAseq_proc/freeze3/5_taxonomy/1_scANVI_psychAD_annotation/230420_PsychAD_MicroPVM_independent_meta_scANVI_new-umap.h5ad"
sce = readH5AD(file, use_hdf5=TRUE, verbose=TRUE, uns=FALSE, layers=FALSE, obsm=FALSE)

counts(sce) = assay(sce, "X")
logcounts(sce) = computeLogCPM(sce)

colData(sce)$dpt_pseudotime = rep(NA, ncol(sce))

idx = dfList[[3]]$idx + 1
colData(sce)$dpt_pseudotime[idx] = dfList[[3]]$dpt_pseudotime

idx = dfList[[4]]$idx + 1
colData(sce)$dpt_pseudotime[idx] = dfList[[4]]$dpt_pseudotime

sce_AH = sce[,sce$subclass %in% c('MG_Adapt', 'MG_Homeo')]
sce_PVM = sce[,sce$subclass %in% c('MG_PVM')]
```

```{r make.plots, eval=FALSE}
df = data.frame(colData(sce_AH), 
	gene = logcounts(sce_AH)["PTPRG",])

df %>%
	filter(gene > 8.1, !is.na(dpt_pseudotime)) %>%
	ggplot(aes(dpt_pseudotime, gene, color=Braak)) +
	geom_point() +
	geom_smooth(method="rlm") + 
	theme_classic() +
	theme(aspect.ratio=1) +
	ylab("PTPRG") +
	ggtitle("Adapt + Homeo")

df = data.frame(colData(sce_PVM), 
	gene = logcounts(sce_PVM)["PTPRG",])

df %>%
	filter(!is.na(dpt_pseudotime)) %>%
	ggplot(aes(dpt_pseudotime, gene, color=Braak)) +
	geom_point() +
	geom_smooth(method="rlm") + 
	theme_classic() +
	theme(aspect.ratio=1) +
	ylab("PTPRG") +
	ggtitle("PVM")
```













<!---

PsychAD metadata
/sc/arion/projects/psychAD/NPS-AD/freeze2_rc/metadata/clinical_metadata.csv
/sc/arion/projects/psychAD/NPS-AD/freeze2_rc/metadata/clinical_metadata_sampleSets_latest.RDS
--->

